components/layout_2020/geom.rs

```diff
@@ -761,3 +774,66 @@ impl LogicalVec2<Size<LengthPercentage>> {
         }
     }
 }
+
+impl Size<Au> {
+    /// Resolves any size into a numerical value.
+    #[inline]
+    pub(crate) fn resolve(
+        &self,
+        get_initial_behavior: impl Fn() -> Self,
+        stretch_size: Au,
+        get_content_size: &mut impl FnMut() -> ContentSizes,
+    ) -> Au {
+        if self.is_initial() {
+            let initial_behavior = get_initial_behavior();
+            assert!(!initial_behavior.is_initial());
+            initial_behavior.resolve_non_initial(stretch_size, get_content_size)
+        } else {
+            self.resolve_non_initial(stretch_size, get_content_size)
+        }
+        .unwrap()
+    }
+
+    /// Resolves a non-initial size into a numerical value.
+    /// Returns `None` if the size is the initial one.
+    #[inline]
+    pub(crate) fn resolve_non_initial(
+        &self,
+        stretch_size: Au,
+        get_content_size: &mut impl FnMut() -> ContentSizes,
+    ) -> Option<Au> {
+        return Some(match self {
+            Self::Keyword(keyword) => match keyword {
+                SizeKeyword::Initial => return None,
+                SizeKeyword::MinContent => get_content_size().min_content,
+                SizeKeyword::MaxContent => get_content_size().max_content,
+                SizeKeyword::FitContent => get_content_size().shrink_to_fit(stretch_size),
+                SizeKeyword::Stretch => stretch_size,
+            },
+            Self::Numeric(numeric) => *numeric,
+        });
```

Comment:
I would go ahead and flatten this as it will be slightly easier to read and shorter:

```suggestion
        match self {
            Self::Keyword(SizeKeyword::Initial) => None,
            Self::Keyword(SizeKeyword::MinContent) => Some(get_content_size().min_content),
            ...
```